operator/builtin/input/azure/eventhub/event_hub.go

```diff
@@ -0,0 +1,202 @@
+		InputOperator: inputOperator,
+		namespace:     c.Namespace,
+		name:          c.Name,
+		group:         c.Group,
+		connStr:       c.ConnectionString,
+		prefetchCount: c.PrefetchCount,
+		startAtEnd:    startAtEnd,
+		persist: Persister{
+			DB: helper.NewScopedDBPersister(buildContext.Database, c.ID()),
+		},
+	}
+	return []operator.Operator{eventHubInput}, nil
+}
+
+// EventHubInput is an operator that reads input from Azure Event Hub.
+type EventHubInput struct {
+	helper.InputOperator
+	cancel context.CancelFunc
+
+	namespace     string
+	name          string
+	group         string
+	connStr       string
+	prefetchCount uint32
+	startAtEnd    bool
+
+	persist Persister
+	hub     *azhub.Hub
+	wg      sync.WaitGroup
+}
+
+// Start will start generating log entries.
+func (e *EventHubInput) Start() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	e.cancel = cancel
+
+	if err := e.persist.DB.Load(); err != nil {
+		return err
+	}
+
+	hub, err := azhub.NewHubFromConnectionString(e.connStr, azhub.HubWithOffsetPersistence(&e.persist))
+	if err != nil {
+		return err
+	}
+	e.hub = hub
+
+	runtimeInfo, err := hub.GetRuntimeInformation(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, partitionID := range runtimeInfo.PartitionIDs {
+		go e.poll(ctx, partitionID, hub)
+	}
+
+	return nil
+}
+
+// Stop will stop generating logs.
+func (e *EventHubInput) Stop() error {
+	e.cancel()
+	e.wg.Wait()
+	if err := e.hub.Close(context.Background()); err != nil {
+		return err
+	}
+	e.Infow(fmt.Sprintf("Closed all connections to Azure Event Hub '%s'", e.name))
+	return e.persist.DB.Sync()
+}
+
+// poll starts polling an Azure Event Hub partition id for new events
+func (e *EventHubInput) poll(ctx context.Context, partitionID string, hub *azhub.Hub) error {
+	offsetStr := ""
+	if e.startAtEnd {
+		offset, err := e.persist.Read(e.namespace, e.name, e.group, partitionID)
+		if err != nil {
+			x := fmt.Sprintf("Error while reading offset for partition_id %s, starting at begining", partitionID)
+			e.Errorw(x, zap.Error(err))
+		} else {
+			offsetStr = offset.Offset
+		}
+	}
+
+	var err error
+
+	// start at begining
+	if !e.startAtEnd {
+		_, err = hub.Receive(
+			ctx, partitionID, e.handleEvent, azhub.ReceiveWithStartingOffset(""),
+			azhub.ReceiveWithPrefetchCount(e.prefetchCount))
+	}
+	// start at end and no offset was found
+	if e.startAtEnd && offsetStr == "" {
+		_, err = hub.Receive(
+			ctx, partitionID, e.handleEvent, azhub.ReceiveWithLatestOffset(),
+			azhub.ReceiveWithPrefetchCount(e.prefetchCount))
+	}
+	// start at end and offset exists
+	if e.startAtEnd && offsetStr != "" {
+		_, err = hub.Receive(
+			ctx, partitionID, e.handleEvent, azhub.ReceiveWithStartingOffset(offsetStr),
+			azhub.ReceiveWithPrefetchCount(e.prefetchCount))
+	}
+	if err != nil {
+		return err
+	}
+
+	e.Infow(fmt.Sprintf("Successfully connected to Azure Event Hub '%s' partition_id '%s'", e.name, partitionID))
+	return nil
```

Comment:
This is well organized, but a bit hard to reason about. I would recommend trying to identify cases where you can exit early and handling those first. Maybe:
```go
// start at begining
if !e.startAtEnd {
	_, err := hub.Receive(
		ctx, partitionID, e.handleEvent, azhub.ReceiveWithStartingOffset(""),
		azhub.ReceiveWithPrefetchCount(e.prefetchCount))
        return err
}
	
// start at end and no offset was found
if offsetStr == "" {
	_, err := hub.Receive(
		ctx, partitionID, e.handleEvent, azhub.ReceiveWithLatestOffset(),
		azhub.ReceiveWithPrefetchCount(e.prefetchCount))
        return err
}
	
// start at end and offset exists
_, err := hub.Receive(
	ctx, partitionID, e.handleEvent, azhub.ReceiveWithStartingOffset(offsetStr),
	azhub.ReceiveWithPrefetchCount(e.prefetchCount))
return err
```

Reply:
In fact, since the first block in the method also is dependent on `startAtEnd`, I would just start the method by checking `if !e.startAtEnd {` and returning right away. Then nothing else has to be concerned with that value, and you can proceed knowing that `startAtEnd` is true.